### PR TITLE
feat: add support for cloning Google Cloud Source Repositories

### PR DESCRIPTION
**Pull request by fengshunli**
Signed-off-by: fsl <1171313930@qq.com>

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [ ] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

### CODE REVIEW

**Comment by crenshaw-dev** [comment]
@fengshunli can you run codegen and push?

**Comment by fengshunli** [comment]
> can you run codegen and push?

@crenshaw-dev There is a test that has failed, please help me find out why

**Comment by crenshaw-dev** [comment]
@fengshunli it's the generated code check. If you run `make codegen` locally, you should see the required changes.

**Comment by fengshunli** [comment]
@crenshaw-dev test still failed

File: util/git/client.go [hunk 1]
```diff
@@ -275,6 +275,18 @@ func newAuth(repoURL string, creds Creds) (transport.AuthMethod, error) {
 		}
 		auth := githttp.BasicAuth{Username: "x-access-token", Password: token}
 		return &auth, nil
+	case GoogleCloudCreds:
+		username, err := creds.getUsername()
+		if err != nil {
+			return nil, err
```
Comment (crenshaw-dev) [inline util/git/client.go:281]:
```suggestion
			return nil, fmt.Errorf("failed to get username from creds: %w", err)
```

File: util/git/client.go [hunk 2]
```diff
@@ -275,6 +275,18 @@ func newAuth(repoURL string, creds Creds) (transport.AuthMethod, error) {
...
+		}
+		token, err := creds.getAccessToken()
+		if err != nil {
+			return nil, err
```
Comment (crenshaw-dev) [inline util/git/client.go:285]:
```suggestion
			return nil, fmt.Errorf("failed to get access token from creds: %w", err)
```

File: util/git/creds.go [hunk 3]
```diff
@@ -424,3 +432,93 @@ func (g GitHubAppCreds) GetClientCertData() string {
...
+		// Invalid JSON
+		log.Errorf("Failed reading credentials from JSON: %+v", err)
+	}
+	return GoogleCloudCreds{creds}
+}
+
+func (c GoogleCloudCreds) Environ() (io.Closer, []string, error) {
+	username, err := c.getUsername()
+	if err != nil {
+		return NopCloser{}, nil, err
```
Comment (crenshaw-dev) [inline util/git/creds.go:453]:
```suggestion
		return NopCloser{}, nil, fmt.Errorf("failed to get username from creds: %w", err)
```

File: util/git/creds.go [hunk 4]
```diff
@@ -424,3 +432,93 @@ func (g GitHubAppCreds) GetClientCertData() string {
...
+	}
+	token, err := c.getAccessToken()
+	if err != nil {
+		return NopCloser{}, nil, err
```
Comment (crenshaw-dev) [inline util/git/creds.go:457]:
```suggestion
		return NopCloser{}, nil, fmt.Errorf("failed to get access token from creds: %w", err)
```

File: util/git/creds.go [hunk 5]
```diff
@@ -424,3 +432,93 @@ func (g GitHubAppCreds) GetClientCertData() string {
...
+		ProjectID    string `json:"project_id"`
+	}
+
+	if c.creds == nil {
+		return "", errors.New("credentials for Google Cloud Source repositories are invalid")
+	}
+
+	var f googleCredentialsFile
+	if err := json.Unmarshal(c.creds.JSON, &f); err != nil {
+		return "", err
```
Comment (crenshaw-dev) [inline util/git/creds.go:484]:
```suggestion
		return "", fmt.Errorf("failed to unmarshal Google Cloud credentials: %w", err)
```

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]

**Comment by alexef** [comment]
superseeded by #11618

**Pull request closed without merging**